File: PHP/dqExpressCustomerClient.php
<?php

require_once('executeModuleInputData.php');
require_once('executeModuleResult.php');

/**
 * This client is used to execute a workflow module
 */
class DqExpressCustomerClient {
	
	const CUSTOMER_SERVICE_MODULE_EXECUTION_URL = "https://<change me>/api/v1/modules/";
	const SESSION_ID_HEADER_NAME = "sessionid";
	
	private $access_token;
	
	public function __construct($token) {
		$this->access_token = $token;
	}

	/** Execute a workflow module.
     * @param moduleName The name of the module to execute
     * @param inputData The module data structure which should be passed to the workflow module
     * @param sessionId The optional session id. We can get a session id by executing a module. 
     * The session id is used to execute similar modules. For example if you call an address suggesst five times with the same address, you only have to pay once.</param>
     * @return Returns the result of the module execution.
     */
	public function executeModule($module_name, $input_data, $session_id) {
		$urlTemplate = DqExpressCustomerClient::CUSTOMER_SERVICE_MODULE_EXECUTION_URL . $module_name;
		
		if(!is_null($session_id) && $session_id !== "") {
			$urlTemplate .= "?sessionId={$session_id}";
		}
		
		$headers = array (
			'Content-Type: application/json',		
			'Authorization: Bearer ' . $this->access_token
		);
		
		$json = json_encode($input_data);
		
		$http = curl_init($urlTemplate);     
		curl_setopt($http, CURLOPT_CUSTOMREQUEST, "PUT");                                                                     
		curl_setopt($http, CURLOPT_POSTFIELDS, $json);                                                                  
		curl_setopt($http, CURLOPT_RETURNTRANSFER, true);     
		curl_setopt($http, CURLOPT_HEADER, true);		
		curl_setopt($http, CURLOPT_HTTPHEADER, $headers); 

		$response = curl_exec ($http);	
		$http_status = curl_getinfo($http, CURLINFO_HTTP_CODE);

		if(curl_errno($http) || $http_status != 200){
			throw new Exception('Something went wrong... Handle your error.');
		}

		$body = $this->getBody($http, $response);
		$headers = $this->getHeaders($http, $response);
		
		curl_close($http);
		
		$json = json_decode($body);
		
		$session_id = "";
		
		if(array_key_exists(DqExpressCustomerClient::SESSION_ID_HEADER_NAME, $headers)) {
			$session_id = $headers[DqExpressCustomerClient::SESSION_ID_HEADER_NAME];
		}
		
		return new ExecuteModuleResult($json, $session_id);
	}
	
	private function getBody($http, $response) {
		$header_size = curl_getinfo($http, CURLINFO_HEADER_SIZE);
		$body = substr($response, $header_size);
		
		return $body;
	}
	
	private function getHeaders($http, $response) {
		$header_size = curl_getinfo($http, CURLINFO_HEADER_SIZE);
		$headerString = substr($response, 0, $header_size);
		$headers=array();

		$data=explode("\n",$headerString);
		$headers['status']=$data[0];

		array_shift($data);

		foreach($data as $part){
			if(!strpos($part, ":")) {
				continue;
			}
			
			$middle=explode(":",$part);
			$headers[strtolower(trim($middle[0]))] = trim($middle[1]);
		}
		
		return $headers;
	}
}

?>

File: C#/DqExpressCustomerClient.cs
using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace DqExpressCustomerClientExample
{
    /// <summary>
    /// This client is used to execute a workflow module
    /// </summary>
    public class DqExpressCustomerClient
    {
        private const string CustomerServiceModuleExecutionUrl = "https://<change me>/api/v1/modules";
        private const string JsonMediaType = "application/json";
        private const string SessionIdHeaderName = "SessionId";

        /// <summary>
        /// Gets or sets the access token which is used for authorization.
        /// </summary>
        public string AccessToken { get; set; }

        /// <summary>
        /// Execute a workflow module.
        /// </summary>
        /// <param name="moduleName">The name of the module to execute.</param>
        /// <param name="inputData">The module data structure which should be passed to the workflow module.</param>
        /// <param name="sessionId">The optional session id. We can get a session id by executing a module. 
        /// The session id is used to execute similar modules. For example if you call an address suggesst five times with the same address, you only have to pay once.</param>
        /// <returns>Returns the result of the module execution.</returns>
        public async Task<ExecuteModuleResult> ExecuteModule(string moduleName,  ExecuteModuleInputData inputData, string sessionId = null)
        {
            var httpClient = new HttpClient();

            httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", AccessToken);

            var url = $"{CustomerServiceModuleExecutionUrl}/{moduleName}";
            if (!string.IsNullOrEmpty(sessionId))
            {
                url += $"?sessionId={sessionId}";
            }

            var stringContent = new StringContent(JsonConvert.SerializeObject(inputData), Encoding.UTF8, JsonMediaType);

            var response = await httpClient.PutAsync(url, stringContent);

            if (response.StatusCode != HttpStatusCode.OK)
            {
                throw new Exception("Something went wrong... Handle your error.");
            }

            var jsonResult = await response.Content.ReadAsStringAsync();
            var result = JsonConvert.DeserializeObject<ExecuteModuleResult>(jsonResult);

            var sessionHeader = response.Headers.FirstOrDefault(header => header.Key.Equals(SessionIdHeaderName, StringComparison.OrdinalIgnoreCase)).Value?.FirstOrDefault();
            if (sessionHeader != null)
            {
                result.SessionId = sessionHeader;
            }

            return result;
        }
    }
}